tutorial/markdown/nodejs/nodejs-langchain-pdf-chat/nodejs-langchain-pdf-chat.md

```diff
@@ -0,0 +1,707 @@
+
+### Read Uploaded PDF File
+
+The code extracts the uploaded file from the request's form data and converts the file to an Array Buffer and then to a Node.js Buffer.
+
+```typescript
+// Get file from form data
+const data = await request.formData();
+const file: File | null = data.get("file") as unknown as File;
+
+// Read file to a Node.js Buffer
+const bytes = await file.arrayBuffer();
+const buffer = Buffer.from(bytes);
+```
+
+### Load PDF File
+
+It then creates a [PDFLoader](<(https://js.langchain.com/docs/integrations/document_loaders/file_loaders/pdf)>) instance from the LangChain library and initializes it with the binary data of the uploaded PDF file. We then load the PDF Files to LangChain [Document](https://js.langchain.com/docs/modules/chains/document/) type.
+
+```typescript
+import { PDFLoader } from "langchain/document_loaders/fs/pdf";
+
+const loader = new PDFLoader(new Blob([buffer]));
+const rawDocs = await loader.load();
+```
+
+### Split Documents
+
+This Langchain document array will contain huge individual files which defeats the purpose while retrieval as we want to send more relevant context to LLM. So we will split it into smaller chunks or passages using LangChain's [_RecursiveCharacterTextSplitter_](https://js.langchain.com/docs/modules/data_connection/document_transformers/recursive_text_splitter):
+
+- chunkSize: 1000: This parameter specifies that each chunk should contain approximately 1000 characters.
+- chunkOverlap: 200: This parameter ensures that there is an overlap of 200 characters between consecutive chunks. This overlap helps maintain context and prevent important information from being split across chunk boundaries.
+
+At the end _splitDocuments_ method splits the large document into smaller Langchain documents based on above defined parameters.
+
+```typescript
+const textSplitter = new RecursiveCharacterTextSplitter({
+  chunkSize: 1000,
+  chunkOverlap: 200,
+});
+const docs = await textSplitter.splitDocuments(rawDocs);
+```
+
+### Initialize OpenAI and Couchbase Vector Store
+
+We will now initialize [OpenAI embeddings](https://js.langchain.com/docs/integrations/text_embedding/openai) which will be used by CouchbaseVectorStore for converting the split docs defined above to vectors (embeddings).
+
+We will also initialize couchbase vector store with couchbase bucket info. Firstly we will connect to couchbase cluster using[`createCouchbaseCluster`](#connecting-to-couchbase).
+
+We will define the bucket, scope, collection and index names from [Environment Variables](#Setup-Environment-Config). We will also define search related parameter:
+
+- `text key` which is the key of where original text will be stored in couchbase.
+- `embedding key` which is the key of where embedded text (vectors) will be stored in couchbase.
```

Comment:
- `embeddingKey` which is the key of where embedded text (vectors) will be stored in couchbase document.